pkg/config/config.go

```diff
@@ -47,16 +47,31 @@ var c config
 var (
 	EnabledEBPFCgroupID = false
 
-	EstimatorModel        = "" // auto-select
-	EstimatorSelectFilter = "" // no filter
-	CoreUsageMetric       = "curr_cpu_cycles"
-	DRAMUsageMetric       = "curr_cache_miss"
-	UncoreUsageMetric     = ""                // no metric (evenly divided)
-	GeneralUsageMetric    = "curr_cpu_cycles" // for uncategorized energy; pkg - core - uncore
+	EstimatorModel        = getConfig("ESTIMATOR_MODEL", "")         // auto-select
+	EstimatorSelectFilter = getConfig("ESTIMATOR_SELECT_FILTER", "") // no filter
+	CoreUsageMetric       = getConfig("CORE_USAGE_METRIC", CPUInstruction)
+	DRAMUsageMetric       = getConfig("DRAM_USAGE_METRIC", CacheMiss)
+	UncoreUsageMetric     = getConfig("UNCORE_USAGE_METRIC", "")              // no metric (evenly divided)
+	GeneralUsageMetric    = getConfig("GENERAL_USAGE_METRIC", CPUInstruction) // for uncategorized energy; pkg - core - uncore
 
 	versionRegex = regexp.MustCompile(`^(\d+)\.(\d+).`)
+
+	ConfigPath = "/etc/config"
 )
 
+func getConfig(key, defaultValue string) string {
```

Comment:
Could you also try to fix these possible "memory leaking" issues?

Reply:
Do you have any guidelines how to fix it?

Reply:
Not sure, but I think you need to check if the variable is null....

Reply:
I think it is the problem of string handling.
https://go101.org/article/memory-leaking.html
Also, there are so many errors in the action log that is not reported...
My guess is that it reported only the first 10 rows. 
<img width="975" alt="Screenshot 2022-10-13 at 09 30 09" src="https://user-images.githubusercontent.com/11749848/195472039-4d6b9ced-9699-4707-99bc-23fdda4d64d0.png">


Why don't we make it as known issue in the code for now and solve it later when we know the exact solution? 
I made multiple attempts but don't want to spam the github action flow for this anymore :sweat_smile:

Reply:
Ok, I investigated the problem a little more.

The scape problem is more related to performance and unnecessary memory allocation.

There are many different contexts where the problem can appear, but here it is related to memory being moved to the heap (scape to heap memory).

This means that after `getConfig` returns, its internal variables are deleted, but the function using `getConfig` will use that variable that no longer exists, and then go will move that value to the heap so that the value is still can be accessed.... More references [here](https://medium.com/a-journey-with-go/go-introduction-to-the-escape-analysis-f7610174e890) and [here](https : //dev.to/human/escape-analysis-in-go-part-1-42bd).

So in practice, most likely the problem is that the `configKey, defaultValue string` values ​​are being used for the `result string` in `getConfig`. However, `configKey, defaultValue` are not being allocated, they are variables or conts.


I think if you create constants it will fix this problem.

```
const (
    	defaultEstimatorModel        =  "ESTIMATOR_MODEL"
	defaultEstimatorSelectFilter = "ESTIMATOR_SELECT_FILTER"
	defaultCoreUsageMetric       = "CORE_USAGE_METRIC"
	defaultDRAMUsageMetric       = "DRAM_USAGE_METRIC"
	defaultUncoreUsageMetric     = "UNCORE_USAGE_METRIC"
	defaultGeneralUsageMetric    = "GENERAL_USAGE_METRIC"
        defaultMetricValue = ""
)

var (
	EnabledEBPFCgroupID = false

	EstimatorModel        = getConfig(defaultEstimatorModel, defaultMetricValue)         // auto-select
	EstimatorSelectFilter = getConfig(defaultEstimatorSelectFilter, defaultMetricValue) // no filter
	CoreUsageMetric       = getConfig(defaultCoreUsageMetric, CPUInstruction)
	DRAMUsageMetric       = getConfig(defaultDRAMUsageMetric, CacheMiss)
	UncoreUsageMetric     = getConfig(defaultUncoreUsageMetric, defaultMetricValue)              // no metric (evenly divided)
	GeneralUsageMetric    = getConfig(defaultGeneralUsageMetric, CPUInstruction) // for uncategorized energy; pkg - core - uncore
)
```

Btw, you can test it locally in your machine and don't need to verify in github...
`make escapes_detect`

Reply:
Tried that in local but does not work...
learking param is still there even I defined those all constants.

Reply:
One significant I pushed before is to use string(([]byte)(configKey)) assigning new variable in the function. 
It removes the leaking param report as shown here. 
```
Error: pkg/config/config.go:63:16: configKey does not escape
Error: pkg/config/config.go:63:27: defaultValue does not escape
```
https://github.com/sustainable-computing-io/kepler/actions/runs/3238779981

However, I don't think it is the point to do so.

Reply:
Right, the problem is because we are using global variable context.....
So it is not creating a local variable which the reference will be use in the `getConfig`

I tried many different solutions without any success and I think your solution is good

```
func getConfig(configKey, defaultValue string) (result string) {
	// making local copies of the global variable parameters to prevent memory leaking
	key := string(([]byte)(configKey))
	result = string(([]byte)(defaultValue))
	configFile := filepath.Join(configPath, key)
	value, err := os.ReadFile(configFile)
	if err == nil {
		result = bytes.NewBuffer(value).String()
	} else {
		strValue, present := os.LookupEnv(key)
		if present {
			result = strValue
		}
	}
	return
}
```

Reply:
Note that the real problem is the `leaking param: configKey`. This is not something that we should ignore...

Another solution is using pointers, and then we will have the message `moved to heap`, which is not optimal but is ok...

Reply:
Got it. Made the changes.